LdapUtils refactor

rvazarkar opened the pull request:
# This MR is not intended to be merged. This is for review purposes

## Description

<!--- Describe your changes in detail -->
Completely refactors connection logic and the API contract. 

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
LdapUtils was a complicated spaghetti mess. This allowed us to fix our API contract, make the library completely async/await supported and clean up our connection logic significantly for compatibility along with fixing several outstanding bugs and implementing new features

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [x] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

rvazarkar committed:
wip: wip

definitelynotagoblin committed:
DC Connection Cache Breakout (#129)

* Breaking dc connection cache out of utils

* Normalize cache key

* Correct domainSid cache key in utils

* Remove unused property Server from LdapConnectionCacheKey

rvazarkar committed:
wip: wip

rvazarkar committed:
wip: remove authtype parameter

rvazarkar committed:
wip: wip

rvazarkar committed:
wip: store off temp result in paged query

rvazarkar committed:
wip: fill out the rest of pagedquery

rvazarkar committed:
wip: implement connection pool

rvazarkar committed:
wip: add connection pool manager

rvazarkar committed:
wip: more wip stuff

rvazarkar committed:
wip: more stuff

rvazarkar committed:
wip: more ldap utils work

rvazarkar committed:
wip: update acl processor

definitelynotagoblin committed:
WIP: rewiring processors to ILdapUtilsNew

rvazarkar committed:
wip: add some missing functions

definitelynotagoblin committed:
Merge branch 'utils_rewrite' into anemeth/utils_rewiring

definitelynotagoblin committed:
LdapPropertyProcessor: use new distinguished name lookup

rvazarkar committed:
wip: processor updates

rvazarkar committed:
wip: more processor fixes

rvazarkar committed:
wip: fix block scoping

definitelynotagoblin committed:
Make some corrections on UserRightsAssignmentProcessor and SearchResultEntryWrapper

rvazarkar committed:
wip: remove some dead code, add some missing functions

rvazarkar committed:
Merge remote-tracking branch 'origin/anemeth/utils_rewiring' into utils_rewrite

rvazarkar committed:
wip: get stuff building

rvazarkar committed:
wip: rename files

rvazarkar committed:
wip: version update

rvazarkar committed:
chore: add test connection

rvazarkar committed:
chore: add disposable to ldap utils

rvazarkar committed:
wip: rename connection wrapper

rvazarkar committed:
wip: more fixes identified during during testing

rvazarkar committed:
wip: fix

rvazarkar committed:
chore: move some code around

rvazarkar committed:
chore: more moving code

definitelynotagoblin committed:
Don't let TestLdapConnection bomb out on exception

definitelynotagoblin committed:
Give up on a domain if NoLdapDataException thrown (or any exception really on CreateLdapConnection)

definitelynotagoblin committed:
Namespace syntax corrections

rvazarkar committed:
wip: comment out tests for initial build

rvazarkar committed:
Merge remote-tracking branch 'origin/utils_rewrite' into utils_rewrite

rvazarkar committed:
wip: comment out more tests

rvazarkar committed:
wip: more test comment

rvazarkar committed:
wip: break more tests

rvazarkar committed:
chore: some test updates

rvazarkar committed:
wip: fix ACLProcessorTest.cs

rvazarkar committed:
chore: fix test

rvazarkar committed:
chore: some pragma fixes

rvazarkar committed:
chore: more mock fixes

rvazarkar committed:
wip: fix computer session tests, mark as available on any platform

rvazarkar committed:
wip: fix container tests

rvazarkar committed:
wip: fix domain trust tests

definitelynotagoblin committed:
Bringing tests back online WIP

rvazarkar committed:
wip: fix gpolocalgroup processor tests

rvazarkar committed:
Merge remote-tracking branch 'origin/utils_rewrite' into utils_rewrite

rvazarkar committed:
wip: fix group processor tests

rvazarkar committed:
wip: uncomment tests

rvazarkar committed:
fix: resolvelabel mistake

definitelynotagoblin committed:
LdapUtilsTest first pass

rvazarkar committed:
wip: ldap utils tests

rvazarkar committed:
IDirectoryObject Refactor (#135)

* wip: implement IDirectoryObject for cleanliness and testability

* wip: move some stuff around, add a bunch of label tests

* chore: add some more tests and fix some subtle bugs

* wip: standardize ldap casing, move producer generation into commonlib

* wip: add docs to ldap utils

* fix: BED-4370 - use lastlogontimestamp in addition to pwdlastset for computer availability

* chore: more tests

* ACL Inheritance Tracking (#133)

* fix: include inheritance id into hash

* feat: add inheritancehashes property to containers/ous/domains

* chore: add tests and fix some old ones

* chore: small update

* chore: add back deleted tests

* wip: patch some tests

* fix: don't try to release a connection if its null

* fix: use value instead of an array reference

* chore: rename entry in wrapper for clarity

* chore: properly use DefaultIfEmpty consistently

* chore: refactor out function for readability

* chore: remove warns

* chore: remove old commented code

rvazarkar committed:
chore: delete backup file

rvazarkar committed:
chore: supress/fix nits

rvazarkar committed:
Replace System.Linq.Async with custom implementations (#137)

* chore: remove system.linq.async, replace with custom functions for file size issues

* chore: add some tests covering new async enumerables

* chore: minor fixes

* chore: add overload for FirstOrDefaultAsync

* chore: add some more tests

rvazarkar committed:
chore: fix incorrect log

rvazarkar committed:
chore: drop log verbosity for reg key opening

rvazarkar committed:
chore: some logging updates

rvazarkar committed:
chore: revert bad change

rvazarkar committed:
chore: cleanups

JonasBK committed:
feat: collector version number meta tag (#139)

rvazarkar committed:
chore: add some missing properties

rvazarkar commented on src/CommonLib/Exceptions/LdapAuthenticationException.cs:
```diff
@@ -5,8 +5,10 @@ namespace SharpHoundCommonLib.Exceptions
 {
     public class LdapAuthenticationException : Exception
```
Make these exceptions private

rvazarkar commented on src/CommonLib/LdapConnectionPool.cs:
```diff
@@ -0,0 +1,371 @@
+    }
+}
+
+//TESTLAB
```
Delete these comments

rvazarkar committed:
chore: visibility changes + remove old comment

urangel commented on test/unit/CommonLibTest.csproj:
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
     <PropertyGroup>
-        <TargetFramework>net5.0</TargetFramework>
```
Suggest updating the dockerfile to also target 7.0

urangel commented on test/unit/ComputerSessionProcessorTest.cs:
```diff
@@ -34,7 +34,7 @@ public void Dispose()
 
         #endregion
 
-        [WindowsOnlyFact]
```
nice updates 📠

urangel commented on test/unit/DomainTrustProcessorTest.cs:
```diff
@@ -53,52 +54,50 @@ public void DomainTrustProcessor_EnumerateDomainTrusts_HappyPath()
                         {"securityIdentifier", Array.Empty<byte>()}
-                    }, "", Label.Domain),
-                new MockSearchResultEntry("CN\u003dexternal.local,CN\u003dSystem,DC\u003dtestlab,DC\u003dlocal",
+                    }, "","")),
```
Is it possible to set default values of empty strings for these last two params in MockDirectoryObject?

rvazarkar replied:
Absolutely

urangel reviewed (approved):
Left some comments but they are not critical

Argentix03 committed:
Update WellKnownPrincipal.cs (#140)

* Update WellKnownPrincipal.cs

The correct type for this is user.  this small thing made a huge confusion for some people today :(
i think this was caused from copy pasting s-1-5-15 the one above it in a fix long ago.

do not confuse IUSR type with IIS_IUSRS
https://learn.microsoft.com/en-us/windows-server/identity/ad-ds/manage/understand-security-identifiers

* Update WellKnownPrincipalTest.cs

rvazarkar committed:
chore: update Dockerfile
chore: set default values for MockDirectoryObject

rvazarkar committed:
Merge remote-tracking branch 'origin/utils_rewrite' into utils_rewrite

rvazarkar committed:
chore: change version to 4.0.0
